### DISCUSSION TITLE
mason equivalent to get_installed_servers()

### DISCUSSION THREAD

**Discussion by WhoIsSethDaniel** [mason-org/mason.nvim#74] [Q&A] [answered]
Currently, for setting up my LSP servers, I do something like this:
```
require'mason'.setup{}
local m = require'mason-lspconfig'
local r = require'mason-registry'
local sm = require "mason-lspconfig.mappings.server"

m.setup {
  ensure_installed = { 'bashls', 'sumneko_lua', 'perlnavigator', 'vimls', 'gopls' },
  automatic_installation = true,
}

local function setup_servers()
  for _, name in pairs(r.get_installed_package_names()) do
    local server = sm.package_to_lspconfig[name]
    require('lspconfig')[server].setup({...config...})
  end
end

setup_servers()
```
That seems fairly involved (at least when compared to `nvim-lsp-installer`. `nvim-lsp-installer` had a very convenient method named `get_installed_servers()`. Are there any thoughts re: having a similar method in Mason? Perhaps a method for each category that does the filtering for you? `get_installed_lsp_servers()`, `get_installed_formatters()`, etc....

Also, currently I can setup `mason-lspconfig` to auto-install LSP servers, but what about non-LSP-servers?  Are there plans to help with auto-installing those?  or is the recommended way to just use `MasonInstall( <tools> )` in our config?

**Comment by williamboman** [comment, marked as answer]
Hello! I've recently added [`:h mason-lspconfig`](https://github.com/williamboman/mason.nvim/blob/alpha/doc/mason-lspconfig.txt) as well as this [reference doc](https://github.com/williamboman/mason.nvim/blob/alpha/doc/reference.md#architecture-diagram) with (hopefully a helpful) diagram.

I'd recommend not accessing `mason-lspconfig.mappings.server` directly as it's not considered public and is subject to breaking changes.

Try something like 

```lua
local m = require("mason-lspconfig")

m.setup {
  ensure_installed = { 'bashls', 'sumneko_lua', 'perlnavigator', 'vimls', 'gopls' },
}

m.setup_handlers {
  function (server_name)
    require('lspconfig')[server_name].setup({...config...})
  end
}
```

> Also, currently I can setup mason-lspconfig to auto-install LSP servers, but what about non-LSP-servers? Are there plans to help with auto-installing those? or is the recommended way to just use MasonInstall( <tools> ) in our config?

Most likely not, as this would wary greatly depending on how you define usage (ideally other plugins would provide this functionality). I'd ideally not have the `mason-lspconfig` extension but seeing how an LSP server is pretty useless by its own (as compared to a linter or formatter), as well as making it easier to transition from nvim-lsp-installer, I figured it'd be worthwhile adding it.

**Reply by WhoIsSethDaniel** [reply]
Thank you. I saw you checking it in and it looks very promising.

On the last question I see at least two use cases:

1. I'm a plugin developer and I want to have a command that installs several third party tools. I want to use Mason to install those tools. I'd like it to be asynchronous and have little user interaction, but I need the user to know if something goes wrong. Ideally I'd also be able to show success as each thing is installed. I may optionally want for the tools to be upgraded, where possible.
2. I'm a user who has a fairly large set of third party tools that I want installed whenever I bring up neovim. If the tools are already installed I may want it to check for updates and update them automatically. Much like 1) I'd like the interaction to be minimal but with some messages printed on success or failure.

These two are fairly similar. Is the best way of dealing with this to develop my own code using the Mason API? (the API mentioned in references.txt, I'm not talking about the user commands). 

Perhaps what I'm asking is: what is the end-state of Mason? Is it strictly an API for installing things (along with some nice convenient user commands and a really nice UI)? Or is more planned?  Either way is fine with me. I have a number of ideas I'd like to try out.

**Reply by williamboman** [reply]
> I'm a plugin developer and I want to have a command that installs several third party tools. I want to use Mason to install those tools. I'd like it to be asynchronous and have little user interaction, but I need the user to know if something goes wrong. Ideally I'd also be able to show success as each thing is installed. I may optionally want for the tools to be upgraded, where possible.

You can achieve this via the [`Package:install({opts})` method](https://github.com/williamboman/mason.nvim/blob/alpha/doc/reference.md#packageinstallopts). It'll return an [`InstallHandle`](https://github.com/williamboman/mason.nvim/blob/alpha/doc/reference.md#installhandle) which you can use to register to different events. The [`Package`](https://github.com/williamboman/mason.nvim/blob/alpha/doc/reference.md#package) instance will also emit events such as `install:success` and `install:failed`. I think this might a bit too low-level for public usage, I've been considering adding a simpler wrapper API.

> I'm a user who has a fairly large set of third party tools that I want installed whenever I bring up neovim. If the tools are already installed I may want it to check for updates and update them automatically. Much like 1) I'd like the interaction to be minimal but with some messages printed on success or failure.

There's a [`Package:check_new_version({callback})`](https://github.com/williamboman/mason.nvim/blob/alpha/doc/reference.md#packagecheck_new_versioncallback) method for this purpose. I've added a note in the description which is worth reading.

> These two are fairly similar. Is the best way of dealing with this to develop my own code using the Mason API? (the API mentioned in references.txt, I'm not talking about the user commands).

Yeah I'd say so! Please go ahead and give it a try - if anything's unclear (which I'm sure there will be) let me know and I can revise the docs further and potentially come up with simpler, more targeted, APIs.

> Perhaps what I'm asking is: what is the end-state of Mason? Is it strictly an API for installing things (along with some nice convenient user commands and a really nice UI)? Or is more planned? Either way is fine with me. I have a number of ideas I'd like to try out.

The primary objective is to be an end-user centric plugin for managing external tooling; in particular LSP & DAP servers, linters, and formatters. The APIs that exist to achieve this are also built to be used for other, custom, use cases - but this is very much a secondary objective for the time being. I'd say it's approximating some kind of end state right now, I've already spent way too much time on this haha. The only thing I see constantly changing going forward is the vendored packages within `mason-registry`.

I'm also keeping an eye on https://github.com/nvim-lua/nvim-package-specification, what has been done here could potentially be extended to implement that spec.

**Comment by MordechaiHadad** [comment]
Hey @williamboman I think you should add this:
```lua
m.setup_handlers {
  function (server_name)
    require('lspconfig')[server_name].setup({...config...})
  end
}
``` 
to readme of mason-lspconfig
I only found about this after searching for similar discussion regarding callback for lsp setup like nvim-lsp-installer had.
